lib/Core/Executor.cpp

```diff
@@ -867,14 +867,58 @@ bool Executor::branchingPermitted(const ExecutionState &state) const {
   return true;
 }
 
+/// Helper method to get an Assignment for a state.  This method is called
+/// when we have reached the maximum number of forks allotted for a particular
+/// exploration.
+Assignment * generateAssignmentForDanglingState(std::map<ExecutionState*, Assignment*> &maxForksMap,
+                                                TimingSolver * solver, ExecutionState &current){
+  Assignment *a = maxForksMap[&current];
+  if (!a && current.symbolics.size() != 0){
+    std::vector< std::vector<unsigned char> > values;
+    std::vector<const Array*> objects;
+    for (unsigned i = 0; i != current.symbolics.size(); ++i){
+      objects.push_back(current.symbolics[i].second);
+    }
+    ConstraintSet extendedConstraints(current.constraints);
+    bool success = solver->getInitialValues(extendedConstraints, objects, values, current.queryMetaData);
+    if (!success) {
+    klee_warning("unable to compute initial values (invalid constraints?)!");
+    ExprPPrinter::printQuery(llvm::errs(), current.constraints,
+                             klee::ConstantExpr::alloc(0, Expr::Bool));
+    return a;
+    }
+
+    a = new Assignment(objects, values);
+    maxForksMap[&current] = a;
+  }
+  return a;
+}
+
 void Executor::branch(ExecutionState &state, 
                       const std::vector< ref<Expr> > &conditions,
                       std::vector<ExecutionState*> &result) {
   TimerStatIncrementer timer(stats::forkTime);
   unsigned N = conditions.size();
   assert(N);
 
-  if (!branchingPermitted(state)) {
+  if (MaxForks!=~0u && stats::forks >= MaxForks) {
+    //See if the state has a previously calculated solution
+    Assignment *a =  generateAssignmentForDanglingState(maxForksMap, solver, state);
+    for (unsigned i=0; i<N; ++i) {
+      ref<Expr> r = a->evaluate(conditions[i]);
+      assert(isa<ConstantExpr>(r) && "Must evaluate to t/f");
+      /// Based on whether the solution follows the true or false branch,
+      /// (Note: it must do one of the two) we add that state to the queue
+      /// to be further extended.  Since we don't want to fork, the other
+      /// option is simply dropped.
+      if (cast<ConstantExpr>(r)->isTrue()){
+        result.push_back(&state);
+      } else {
+        result.push_back(nullptr);
+      }
+    }
+  }
+  else if (!branchingPermitted(state)) {
```

Comment:
`branchingPermitted` tests for the same condition. Having 4 different conditions that prevent forking handled in 2 different ways is not very intuitive. IMHO this should be re-factored into different functions with proper naming instead of repeating `MaxForks!=~0u && stats::forks >= MaxForks` everywhere and increasing the amount of spaghetti code in the Executor.

Reply:
I add a new function `reachMaxforks` to judge whether reach max fork limit

Reply:
BTW, `branchingPermitted` should also call this function.